lib/rust/flexer/src/generate.rs

```diff
@@ -0,0 +1,531 @@
+    let mut dfa_transitions = transitions_for_dfa(&mut dfa,group.id.into())?;
+    dfa_transitions.insert(0,dispatch_for_dfa);
+    dfa_transitions.extend(rules);
+    Ok(dfa_transitions)
+}
+
+/// Generate a set of transition functions for the provided `dfa`, with identifier `id`.
+pub fn transitions_for_dfa(dfa:&mut DFA, id:usize) -> Result<Vec<ImplItem>,GenError> {
+    let mut state_has_overlapping_rules:HashMap<usize,bool> = HashMap::new();
+    state_has_overlapping_rules.insert(0,false);
+    let state_names:Vec<_> = dfa.links.row_indices().map(|ix| (ix, name_for_step(id, ix))).collect();
+    let mut transitions    = Vec::with_capacity(state_names.len());
+    for (ix,name) in state_names.into_iter() {
+        transitions.push(transition_for_dfa(dfa,name,ix,&mut state_has_overlapping_rules)?)
+    }
+    Ok(transitions)
+}
+
+/// Generate a specific transition function for
+#[allow(clippy::implicit_hasher)]
+pub fn transition_for_dfa<S:BuildHasher>
+( dfa:&mut DFA
+, transition_name:Ident
+, state_ix:usize
+, has_overlaps:&mut HashMap<usize,bool,S>
+) -> Result<ImplItem,GenError> {
+    let match_expr:Expr   = match_for_transition(dfa,state_ix,has_overlaps)?;
+    let function:ImplItem = parse_quote! {
+        fn #transition_name<R:LazyReader>(&mut self, reader:&mut R) -> StageStatus {
+            #match_expr
+        }
+    };
+    Ok(function)
+}
+
+/// Generate the pattern match for a given transition function.
+pub fn match_for_transition<S:BuildHasher>
+( dfa:&mut DFA
+, state_ix:usize
+, has_overlaps:&mut HashMap<usize,bool,S>
+) -> Result<Expr,GenError> {
+    let overlaps          = *has_overlaps.get(&state_ix).unwrap_or(&false);
+    let state             = dfa.callbacks.get(state_ix).expect("Internal error.").clone();
+    let mut trigger_state = dfa.links[(state_ix,0)];
+    let mut range_start   = u32::min_value();
+    let divisions:Vec<_>  = dfa.alphabet_segmentation.divisions_as_vec();
+    let mut branches      = Vec::with_capacity(divisions.len());
+    for (ix,sym) in divisions.into_iter() {
+        let new_trigger_state = dfa.links[(state_ix,ix)];
+        if new_trigger_state != trigger_state {
+            let range_end             = if sym.val != 0 { sym.val - 1 } else { sym.val };
+            let current_trigger_state = trigger_state;
+            let current_range_start   = range_start;
+            trigger_state             = new_trigger_state;
+            range_start               = sym.val;
+            let body =
+                branch_body(dfa,current_trigger_state,&state,has_overlaps,overlaps)?;
+            branches.push(Branch::new(Some(current_range_start..=range_end),body))
+        } else {}
+    }
+    let catch_all_branch_body = branch_body(dfa,trigger_state,&state,has_overlaps,overlaps)?;
+    let catch_all_branch      = Branch::new(None,catch_all_branch_body);
+    branches.push(catch_all_branch);
+    let arms:Vec<Arm> = branches.into_iter().map(|b| b.into()).collect();
```

Comment:
`map(|b| b.into())` can be written without introducing a name, as `map(Into::into)`